src/App.tsx

```diff
@@ -1,18 +1,95 @@
 import React from 'react';
-import { Routes, Route } from 'react-router-dom';
+import { Route, Routes, useLocation, useNavigate } from 'react-router-dom';
+import Exercises from './pages/Exercises';
+import Home from './pages/Home';
 import Layout from './pages/Layout';
 import NotFound from './pages/NotFound';
-import Home from './pages/Home';
+import Profile from './pages/Profile';
+import SignIn from './pages/SignIn';
+import SignUp from './pages/SignUp';
+import Unauthorized from './pages/Unauthorized';
+import { AuthenticatedUser } from './store/userSlice';
+import useAuth from './utils/useAuth';
+
+interface RouteConfig {
+  element: React.ReactNode;
+  path: string;
+  authHandler: (auth: AuthenticatedUser | null) => boolean;
+}
+
+function ProtectedRoutes({ routes }: { routes: Array<RouteConfig> }) {
```

Comment:
Right now your ProtectedRoutes logic doesn't seem to be working... I am able to access both `/profile` and `/exercises` when logged out. The Firebase sign-up/sign-in does appear to be working correctly, though.

 I'd advise you to get the route protection working so only signed-in users can access those routes.